src/lib.rs

```diff
@@ -1644,4 +1652,463 @@ mod tests {
+          &is_left_child_smaller,
+        )?;
+
+        lookup_trace.write(
+          cs.namespace(|| "write_left_pair_parent"),
+          index,
+          &new_parent_left,
+        )?;
+        lookup_trace.write(
+          cs.namespace(|| "write_left_pair_child"),
+          &left_child_index,
+          &new_left_child,
+        )?;
+
+        let is_right_child_smaller = less_than(
+          cs.namespace(|| "right_child < parent"),
+          &right_child,
+          &new_parent_left,
+          self.max_value_bits,
+        )?;
+
+        let new_parent_right = conditionally_select2(
+          cs.namespace(|| "new_right_pair_parent"),
+          &right_child,
+          &new_parent_left,
+          &is_right_child_smaller,
+        )?;
+        let new_right_child = conditionally_select2(
+          cs.namespace(|| "new_right_pair_child"),
+          &new_parent_left,
+          &right_child,
+          &is_right_child_smaller,
+        )?;
+        lookup_trace.write(
+          cs.namespace(|| "write_right_pair_parent"),
+          index,
+          &new_parent_right,
+        )?;
+        lookup_trace.write(
+          cs.namespace(|| "write_left_pair_child"),
+          &right_child_index,
+          &new_right_child,
+        )?;
+
+        // commit the rw change
+        let (next_R, next_W, next_rw_counter, next_intermediate_gamma) = lookup_trace
+          .commit::<G2, Namespace<'_, F, <CS as ConstraintSystem<F>>::Root>>(
+            cs.namespace(|| "commit"),
+            self.ro_consts.clone(),
+            prev_intermediate_gamma,
+            &(alpha.clone(), gamma.clone()),
+            prev_W,
+            prev_R,
+            prev_rw_counter,
+          )?;
+
+        let next_index = AllocatedNum::alloc(cs.namespace(|| "next_index"), || {
+          index
+            .get_value()
+            .map(|index| index - G1::Scalar::from(1))
+            .ok_or(SynthesisError::AssignmentMissing)
+        })?;
+        cs.enforce(
+          || "(next_index + 1) * 1 = index",
+          |lc| lc + next_index.get_variable() + CS::one(),
+          |lc| lc + CS::one(),
+          |lc| lc + index.get_variable(),
+        );
+        Ok(vec![
+          next_intermediate_gamma,
+          alpha.clone(),
+          gamma.clone(),
+          next_R,
+          next_W,
+          next_rw_counter,
+          next_index,
+        ])
+      }
+    }
+
+    /// A trivial step circuit that simply returns the input
+    #[derive(Clone, Debug, Default, PartialEq, Eq)]
+    pub struct TrivialTestCircuit<F: PrimeField> {
+      _p: PhantomData<F>,
+    }
+
+    impl<F> StepCircuit<F> for TrivialTestCircuit<F>
+    where
+      F: PrimeField,
+    {
+      fn arity(&self) -> usize {
+        1
+      }
+
+      fn synthesize<CS: ConstraintSystem<F>>(
+        &self,
+        _cs: &mut CS,
+        z: &[AllocatedNum<F>],
+      ) -> Result<Vec<AllocatedNum<F>>, SynthesisError> {
+        Ok(z.to_vec())
+      }
+    }
+
```

Comment:
Consider reusing, rather than re-defining, `traits::circuit::TrivialCircuit<F>`

Reply:
Fixed, It just some left over when previously when experimenting mutable `stepcircuit`